## 4.9 Intro to Data Visualizations with Python Part 1

# 1. Importing libraries

In [2]:
# Import libraries
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

# 1.1 Importing Data

In [3]:
path = r'C:\Users\Gregor\ownCloud\CareerFoundry\10-2023 Instacart Basket Analysis'

In [4]:
# Import the orders_checked.csv file
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'), index_col = False)

In [5]:
# Import the products_checked.csv file
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'), index_col = False)

In [6]:
# Import the orders_products_aggregated.pkl file
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_aggregated.pkl'))

## Step 3: Import your analysis libraries, as well as your new customer data set as a dataframe.

In [11]:
# Import the customers.csv file
customer = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'), index_col = False)

## Step 4: Wrangle the data so that it follows consistent logic; for example, rename columns with illogical names and drop columns that don’t add anything to your analysis.

In [12]:
customer

user_id First Name    Surnam  Gender           STATE  Age date_joined  \
0         26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1         33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2         65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3        125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4        130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
...         ...        ...       ...     ...             ...  ...         ...   
206204   168073       Lisa      Case  Female  North Carolina   44    4/1/2020   
206205    49635     Jeremy   Robbins    Male          Hawaii   62    4/1/2020   
206206   135902      Doris  Richmond  Female        Missouri   66    4/1/2020   
206207    81095       Rose   Rollins  Female      California   27    4/1/2020   
206208    80148    Cynthia     Noble  Female        New York   55    4/1/2020   

        n_dependants fam_status  income  
0                  3    married  165665  
1                  0     single   59285  
2                  2    married   99568  
3                  0     single   42049  
4                  1    married   40374  
...              ...        ...     ...  
206204             1    married  148828  
206205             3    married  168639  
206206             2    married   53374  
206207             1    married   99799  
206208             1    married   57095  

[206209 rows x 10 columns]

In order to stay consistent with the other dataframes I rename the columns to lowercase and replace spaces with underscores.

In [13]:
# changing the name of columns 
customer.rename(columns = {'First Name' : 'first_name'}, inplace = True)
customer.rename(columns = {'Surnam' : 'surname'}, inplace = True)
customer.rename(columns = {'Gender' : 'gender'}, inplace = True)
customer.rename(columns = {'STATE': 'state'}, inplace = True)
customer.rename(columns = {'Age' : 'age'}, inplace = True)
customer.rename(columns = {'n_dependants' : 'number_of_dependants'}, inplace = True)
customer.rename(columns = {'fam_status' : 'marital_status'}, inplace = True)

In [14]:
# Check
customer

user_id first_name   surname  gender           state  age date_joined  \
0         26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1         33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2         65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3        125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4        130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
...         ...        ...       ...     ...             ...  ...         ...   
206204   168073       Lisa      Case  Female  North Carolina   44    4/1/2020   
206205    49635     Jeremy   Robbins    Male          Hawaii   62    4/1/2020   
206206   135902      Doris  Richmond  Female        Missouri   66    4/1/2020   
206207    81095       Rose   Rollins  Female      California   27    4/1/2020   
206208    80148    Cynthia     Noble  Female        New York   55    4/1/2020   

        number_of_dependants marital_status  income  
0                          3        married  165665  
1                          0         single   59285  
2                          2        married   99568  
3                          0         single   42049  
4                          1        married   40374  
...                      ...            ...     ...  
206204                     1        married  148828  
206205                     3        married  168639  
206206                     2        married   53374  
206207                     1        married   99799  
206208                     1        married   57095  

[206209 rows x 10 columns]

## Step 5: Complete the fundamental data quality and consistency checks you’ve learned throughout this Achievement; for example, check for and address missing values and duplicates, and convert any mixed-type data.

In [15]:
# Checking for missing values and summing them

customer.isnull().sum()

user_id                     0
first_name              11259
surname                     0
gender                      0
state                       0
age                         0
date_joined                 0
number_of_dependants        0
marital_status              0
income                      0
dtype: int64

There are 11259 missing values in the first_name column.

In [18]:
# Checking for duplicates

customer[customer.duplicated()]

Empty DataFrame
Columns: [user_id, first_name, surname, gender, state, age, date_joined, number_of_dependants, marital_status, income]
Index: []

There are no duplicates.

In [16]:
# Checking to see if there are any mixed types
for col in customer.columns.tolist():
    weird = (customer[[col]].applymap(type) != customer[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (customer[weird]) > 0:
        print (col)

C:\Users\Gregor\AppData\Local\Temp\ipykernel_16436\1902448255.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (customer[[col]].applymap(type) != customer[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Gregor\AppData\Local\Temp\ipykernel_16436\1902448255.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (customer[[col]].applymap(type) != customer[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Gregor\AppData\Local\Temp\ipykernel_16436\1902448255.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (customer[[col]].applymap(type) != customer[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Gregor\AppData\Local\Temp\ipykernel_16436\1902448255.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (customer[[col]].applymap(type) != customer[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Gregor\

first_name


C:\Users\Gregor\AppData\Local\Temp\ipykernel_16436\1902448255.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (customer[[col]].applymap(type) != customer[[col]].iloc[0].apply(type)).any(axis = 1)


The column first_name has mixed types which is probably due to the missing values.

In [19]:
# Looking at the missing values
customer.loc[customer['first_name'].isna()]

user_id first_name   surname  gender         state  age date_joined  \
53        76659        NaN   Gilbert    Male      Colorado   26    1/1/2017   
73        13738        NaN     Frost  Female     Louisiana   39    1/1/2017   
82        89996        NaN    Dawson  Female        Oregon   52    1/1/2017   
99        96166        NaN   Oconnor    Male      Oklahoma   51    1/1/2017   
105       29778        NaN    Dawson  Female          Utah   63    1/1/2017   
...         ...        ...       ...     ...           ...  ...         ...   
206038   121317        NaN    Melton    Male  Pennsylvania   28   3/31/2020   
206044   200799        NaN  Copeland  Female        Hawaii   52    4/1/2020   
206090   167394        NaN     Frost  Female        Hawaii   61    4/1/2020   
206162   187532        NaN     Floyd  Female    California   39    4/1/2020   
206171   116898        NaN   Delgado  Female      Colorado   23    4/1/2020   

        number_of_dependants marital_status  income  
53                         2        married   41709  
73                         0         single   82518  
82                         3        married  117099  
99                         1        married  155673  
105                        3        married  151819  
...                      ...            ...     ...  
206038                     3        married   87783  
206044                     2        married  108488  
206090                     1        married   45275  
206162                     0         single   56325  
206171                     2        married   59222  

[11259 rows x 10 columns]

In [20]:
# Addressing missing values by imputation
customer['first_name'].fillna('Unknown', inplace = True)

In [21]:
# Rechecking missing values
customer.isnull().sum() 

user_id                 0
first_name              0
surname                 0
gender                  0
state                   0
age                     0
date_joined             0
number_of_dependants    0
marital_status          0
income                  0
dtype: int64

In [22]:
# Rechecking mixed types
for col in customer.columns.tolist():
    weird = (customer[[col]].applymap(type) != customer[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (customer[weird]) > 0:
        print (col)

C:\Users\Gregor\AppData\Local\Temp\ipykernel_16436\3510070292.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (customer[[col]].applymap(type) != customer[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Gregor\AppData\Local\Temp\ipykernel_16436\3510070292.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (customer[[col]].applymap(type) != customer[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Gregor\AppData\Local\Temp\ipykernel_16436\3510070292.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (customer[[col]].applymap(type) != customer[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Gregor\AppData\Local\Temp\ipykernel_16436\3510070292.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (customer[[col]].applymap(type) != customer[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Gregor\

After replacing the missing values in the first_name column, there are no mixed types and to missing values anymore.

## Step 6: Combine your customer data with the rest of your prepared Instacart data. (Hint: Make sure the key columns are the same data type!)

In [24]:
ords_prods_merge.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32404859 entries, 0 to 32404858
Data columns (total 25 columns):
 #   Column                  Dtype   
---  ------                  -----   
 0   order_id                int64   
 1   user_id                 int64   
 2   eval_set                object  
 3   order_number            int64   
 4   orders_day_of_week      int64   
 5   order_hour_of_day       int64   
 6   days_since_prior_order  float64 
 7   product_id              int64   
 8   add_to_cart_order       int64   
 9   reordered               int64   
 10  product_name            object  
 11  aisle_id                int64   
 12  department_id           int64   
 13  prices                  float64 
 14  _merge                  category
 15  price_range_loc         object  
 16  busiest_day             object  
 17  busiest_days            object  
 18  busiest_period_of_day   object  
 19  max_order               int64   
 20  loyalty_flag            object  
 21  avg_pr

In [25]:
customer.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column                Non-Null Count   Dtype 
---  ------                --------------   ----- 
 0   user_id               206209 non-null  int64 
 1   first_name            206209 non-null  object
 2   surname               206209 non-null  object
 3   gender                206209 non-null  object
 4   state                 206209 non-null  object
 5   age                   206209 non-null  int64 
 6   date_joined           206209 non-null  object
 7   number_of_dependants  206209 non-null  int64 
 8   marital_status        206209 non-null  object
 9   income                206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


In [26]:
# Merging ords_prods_merge with customer df on user_id, as user_id has in both cases the data type int64
ords_prods_customer_merge = ords_prods_merge.merge(customer, on = ['user_id'])

In [27]:
# Checking final output
ords_prods_customer_merge

order_id  user_id eval_set  order_number  orders_day_of_week  \
0          2539329        1    prior             1                   2   
1          2398795        1    prior             2                   3   
2           473747        1    prior             3                   3   
3          2254736        1    prior             4                   4   
4           431534        1    prior             5                   4   
...            ...      ...      ...           ...                 ...   
32404854    156685   106143    prior            26                   4   
32404855    484769    66343    prior             1                   6   
32404856   1561557    66343    prior             2                   1   
32404857    276317    66343    prior             3                   6   
32404858   2922475    66343    prior             4                   1   

          order_hour_of_day  days_since_prior_order  product_id  \
0                         8                     NaN         196   
1                         7                    15.0         196   
2                        12                    21.0         196   
3                         7                    29.0         196   
4                        15                    28.0         196   
...                     ...                     ...         ...   
32404854                 23                     5.0       19675   
32404855                 11                     NaN       47210   
32404856                 11                    30.0       47210   
32404857                 15                    19.0       47210   
32404858                 12                    30.0       47210   

          add_to_cart_order  reordered  ...         frequency_flag  \
0                         1          0  ...  Non-frequent customer   
1                         1          1  ...  Non-frequent customer   
2                         1          1  ...  Non-frequent customer   
3                         1          1  ...  Non-frequent customer   
4                         1          1  ...  Non-frequent customer   
...                     ...        ...  ...                    ...   
32404854                  1          1  ...      Frequent customer   
32404855                  1          0  ...  Non-frequent customer   
32404856                  1          1  ...  Non-frequent customer   
32404857                  1          1  ...  Non-frequent customer   
32404858                  1          1  ...  Non-frequent customer   

          first_name  surname  gender      state age date_joined  \
0              Linda   Nguyen  Female    Alabama  31   2/17/2019   
1              Linda   Nguyen  Female    Alabama  31   2/17/2019   
2              Linda   Nguyen  Female    Alabama  31   2/17/2019   
3              Linda   Nguyen  Female    Alabama  31   2/17/2019   
4              Linda   Nguyen  Female    Alabama  31   2/17/2019   
...              ...      ...     ...        ...  ..         ...   
32404854      Gerald    Yates    Male     Hawaii  25   5/26/2017   
32404855  Jacqueline   Arroyo  Female  Tennessee  22   9/12/2017   
32404856  Jacqueline   Arroyo  Female  Tennessee  22   9/12/2017   
32404857  Jacqueline   Arroyo  Female  Tennessee  22   9/12/2017   
32404858  Jacqueline   Arroyo  Female  Tennessee  22   9/12/2017   

         number_of_dependants marital_status  income  
0                           3        married   40423  
1                           3        married   40423  
2                           3        married   40423  
3                           3        married   40423  
4                           3        married   40423  
...                       ...            ...     ...  
32404854                    0         single   53755  
32404855                    3        married   46151  
32404856                    3        married   46151  
32404857                    3        married   46151  
32404858                    3        married   46151 

In [28]:
ords_prods_customer_merge.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32404859 entries, 0 to 32404858
Data columns (total 34 columns):
 #   Column                  Dtype   
---  ------                  -----   
 0   order_id                int64   
 1   user_id                 int64   
 2   eval_set                object  
 3   order_number            int64   
 4   orders_day_of_week      int64   
 5   order_hour_of_day       int64   
 6   days_since_prior_order  float64 
 7   product_id              int64   
 8   add_to_cart_order       int64   
 9   reordered               int64   
 10  product_name            object  
 11  aisle_id                int64   
 12  department_id           int64   
 13  prices                  float64 
 14  _merge                  category
 15  price_range_loc         object  
 16  busiest_day             object  
 17  busiest_days            object  
 18  busiest_period_of_day   object  
 19  max_order               int64   
 20  loyalty_flag            object  
 21  avg_pr

## Step 8: Export this new dataframe as a pickle file so you can continue to use it in the second part of this task.

In [ ]:
# Export ords_prods_customer_merge as a pickle file 
ords_prods_customer_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_customer_merge.pkl'))